### ISSUE TITLE
3.0 test userconfig types export ok

### ISSUE THREAD

**Issue by zachleat** [11ty/eleventy#3097]
(This one is just a reminder for me)

See: https://www.11ty.dev/docs/config/#type-definitions

Pre-alpha I can confirm that v9 (ESM) of `eleventy-base-blog` with npm installed via `file:../eleventy` works fine but I want to remind myself to check post-publish too.

**Comment by rubenwardy** [comment]
This doesn't work for me:

```js
import { UserConfig } from "@11ty/eleventy";
```

I need to do:

```js
import UserConfig from "@11ty/eleventy/src/UserConfig.js";
```

------------------------

```
[11ty] 1. Error in your Eleventy config file '.eleventy.js'. (via EleventyConfigError)
[11ty] 2. The requested module '@11ty/eleventy' does not provide an export named 'UserConfig' (via SyntaxError)
[11ty] 
[11ty] Original error stack trace: file:///home/ruben/dev/web/blog/plugins/wordcount.js:1
[11ty] import {UserConfig} from "@11ty/eleventy";
[11ty]         ^^^^^^^^^^
[11ty] SyntaxError: The requested module '@11ty/eleventy' does not provide an export named 'UserConfig'
[11ty]     at ModuleJob._instantiate (node:internal/modules/esm/module_job:124:21)
[11ty]     at async ModuleJob.run (node:internal/modules/esm/module_job:190:5)
```

**Comment by rubenwardy** [comment]
see also https://github.com/11ty/eleventy/issues/3127

**Comment by uncenter** [comment]
#3060 removed the .d.ts file to rely on JSDoc entirely, but I think the issue is UserConfig isn't exported anywhere else now. Would the fix be to export it [here](https://github.com/11ty/eleventy/blob/main/src/Eleventy.js#L1291-L1305)?

**Comment by what1s1ove** [comment]
> #3060 removed the .d.ts file to rely on JSDoc entirely, but the .d.ts had been exporting UserConfig (https://github.com/11ty/eleventy/blob/ecd0579a2dded6939510b7c23841388b26eb6d16/src/index.d.ts) and the issue now that is UserConfig isn't exported anywhere else. Would the fix be to export it [here](https://github.com/11ty/eleventy/blob/main/src/Eleventy.js#L1291-L1305)?

Agree. I have `tsconfig` which checks all JS files (including `eleventy.config.js`). Now I need to manually describe all the config keys 😕

**Comment by uncenter** [comment]
> > #3060 removed the .d.ts file to rely on JSDoc entirely, but the .d.ts had been exporting UserConfig ([`ecd0579`/src/index.d.ts](https://github.com/11ty/eleventy/blob/ecd0579a2dded6939510b7c23841388b26eb6d16/src/index.d.ts)) and the issue now that is UserConfig isn't exported anywhere else. Would the fix be to export it [here](https://github.com/11ty/eleventy/blob/main/src/Eleventy.js?rgh-link-date=2023-12-23T15%3A02%3A03Z#L1291-L1305)?
> 
> Agree. I have `tsconfig` which checks all JS files (including `eleventy.config.js`). Now I need to manually describe all the config keys 😕

You don't have to do that, can't you just use it like `/** @param { import("@11ty/eleventy/src/UserConfig.js") } eleventyConfig */` (as suggested [here](https://github.com/11ty/eleventy/issues/3127#issue-2048464746))?

**Comment by what1s1ove** [comment]
> > > #3060 removed the .d.ts file to rely on JSDoc entirely, but the .d.ts had been exporting UserConfig ([`ecd0579`/src/index.d.ts](https://github.com/11ty/eleventy/blob/ecd0579a2dded6939510b7c23841388b26eb6d16/src/index.d.ts)) and the issue now that is UserConfig isn't exported anywhere else. Would the fix be to export it [here](https://github.com/11ty/eleventy/blob/main/src/Eleventy.js?rgh-link-date=2023-12-23T15%3A02%3A03Z#L1291-L1305)?
> > 
> > 
> > Agree. I have `tsconfig` which checks all JS files (including `eleventy.config.js`). Now I need to manually describe all the config keys 😕
> 
> You don't have to do that, can't you just use it like `/** @param { import("@11ty/eleventy/src/UserConfig.js") } eleventyConfig */` (as suggested [here](https://github.com/11ty/eleventy/issues/3127#issue-2048464746))?

Unfortunately, no...

Today I migrated to a new version. I'm so happy that the number of cjs files is decreasing in my projects 🎉
https://github.com/what1s1ove/whatislove.dev/pull/279

After migration, I encountered two issues.

1. "Replace await `readFile(pathTojson)` with Node.js import attributes (versions 18 and 20 are different!)". Discussed in this issue - https://github.com/11ty/eleventy/issues/3128
  - https://nodejs.org/docs/latest-v18.x/api/esm.html#import-attributes
  - https://nodejs.org/docs/latest-v20.x/api/esm.html#import-attributes
2. "Replace custom type definitions for the 11ty package" – relates to this issue.

As I mentioned earlier, my project is configured so that TypeScript performs checks on all `.js` files (in fact, not only TypeScript but many other linters do the same). Before migrating the 11ty config to js, it was ignored because it had the `.cjs` extension.

When TypeScript encounters an import of a package that has no types, it suggests two solutions:

1. Install types separately. In our case, this is not possible, as there are no types for 11ty as a separate package (and there's no need to have them as a separate package 🤞).
2. Create custom types using `declare module`. However, the problem with `declare module` is that within this construct, you cannot import packages that don't have types (well, you can, but any such import would result in the type any). Because of this, all types should be defined manually.

<img width="1084" alt="image" src="https://github.com/11ty/eleventy/assets/42296182/26bb7006-1cbb-4d9c-ae39-4f4e9f4a24d4">

I think before migrating to ESM, this was a rare problem, as I mentioned, many linters ignored `.cjs` by default. Because of this, the 11ty config didn't fall under various linters (eslint, TypeScript, etc). But now, since this will be the default behavior, this problem will affect many users. It seems to me that it should be addressed by adding `d.ts` files to the 11ty package itself.

**Comment by murtuzaalisurti** [comment]
While migrating to  [11tyv3.0.0-alpha.5](https://github.com/11ty/eleventy/tree/v3.0.0-alpha.5), the JSDoc comments stopped working for me. It seems that in ESM, you have to explicitly specify the default export.

```js
/** @param { import("@11ty/eleventy").UserConfig } eleventyConfig */ ❌

/** @param { (import("@11ty/eleventy/src/UserConfig").default) } eleventyConfig */ ✅
```

**Comment by zachleat** [comment]
See #3291

**Comment by zachleat** [comment]
`/** @param {import("@11ty/eleventy").UserConfig} eleventyConfig */` seems to work with 3.0.0-alpha.12 (shipped by #3291).

Closing this but would appreciate further confirmation!

**Comment by what1s1ove** [comment]
Hey guys!

Does this work for anyone? It finds the import, but it is marked as `any`.

<img width="535" alt="image" src="https://github.com/11ty/eleventy/assets/42296182/6156ee73-62e1-4511-908c-d40d0b5135a0">

**Comment by zachleat** [comment]
@mayank99 helped out here with #3349, shipping with 3.0.0-alpha.15

**Comment by what1s1ove** [comment]
> @mayank99 helped out here with #3349, shipping with 3.0.0-alpha.15

Hey @zachleat ! Could you let me know when you're going to publish it?

**Comment by tbroyer** [comment]
I have `@param {import("@11ty/eleventy").UserConfig} eleventyConfig` in JSDoc that works with 3.0.0-alpha.14 but fails with 3.0.0-alpha.16 with the same error as in https://github.com/11ty/eleventy/issues/3097#issuecomment-1868322883 when using `tsc` to lint my JS code.

```
.eleventy.js:22:19 - error TS7016: Could not find a declaration file for module '@11ty/eleventy'. '…/node_modules/@11ty/eleventy/src/Eleventy.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/11ty__eleventy` if it exists or add a new declaration (.d.ts) file containing `declare module '@11ty/eleventy';`

22  * @param {import("@11ty/eleventy").UserConfig} eleventyConfig
                     ~~~~~~~~~~~~~~~~
```

**Comment by mayank99** [comment]
@tbroyer I just tested `3.0.0-alpha.16` and this works totally fine.

```js
/** @param {import("@11ty/eleventy").UserConfig} eleventyConfig */
export default function (eleventyConfig) {
```

Here's a minimal stackblitz link to confirm: https://stackblitz.com/edit/github-7xpr1h?file=.eleventy.js

Maybe you can share a minimal repro if it's not working for you? It's likely that some custom tsconfig rules might be interfering with it.

**Comment by tbroyer** [comment]
This stackblitz actually exhibits that same error:
![image](https://github.com/11ty/eleventy/assets/173813/394b3d38-aa7b-45af-9412-c296f65b9dea)

**Comment by tbroyer** [comment]
Got it: this is `tsc`'s `--strict`, that generates an error out of this: https://stackblitz.com/edit/github-7xpr1h-crqrvo?file=.eleventy.js,package.json

Because 3.0.0-alpha.14 had a `"type"` entry in its `package.json` pointing to a `.d.ts`, this worked great. We already had to add `@ts-ignore` to a couple `import`s like I did with `@11ty/eleventy-plugin-webc` here, but at least it worked with `@11ty/eleventy`.

Should a `.d.ts` be generated out of the JSDoc comments? (and added back to the `package.json`)

**Comment by zachleat** [comment]
Spent some time looking into this and found a few things worth noting:

v3.0.0-alpha.14 failed `tsc --strict` with the following error:

```
~/Temp/eleventy-3097-ts-strict ᐅ npx tsc --noEmit --checkJs --allowJS --strict .eleventy.js
node_modules/@11ty/eleventy/src/index.d.ts:1:24 - error TS7016: Could not find a declaration file for module './UserConfig'. '/Users/zachleat/Temp/eleventy-3097-ts-strict/node_modules/@11ty/eleventy/src/UserConfig.js' implicitly has an 'any' type.

1 import UserConfig from "./UserConfig";
                         ~~~~~~~~~~~~~~


Found 1 error in node_modules/@11ty/eleventy/src/index.d.ts:1
```

In v3.0.0-alpha.14 `tsc` worked without `--strict`.

Irrespective of `--strict`, the "viral" nature of TypeScript requires the Eleventy codebase to obey TypeScript conventions for any types we want to export. I attempted to use the `noResolve` to work around this and only provide types for `Eleventy` and `UserConfig` (skipping a large refactor) but this caused other issues with node types: https://fediverse.zachleat.com/@zachleat/112848095569188008

The `@typedef` approach we swapped to here seems to be causing issues with `tsc` because it runs through Eleventy.js and checks it and all of Eleventy.js’ dependencies (lots of errors) on the way to UserConfig.js.

The previous approach worked better only because it exported `UserConfig` directly (not via the `Eleventy` object), e.g. `index.d.ts`:

```js
import UserConfig from "./src/UserConfig.js";

export { UserConfig };
```

That said, I think it’d be useful to have IDE autocomplete and `tsc` compatibility for `Eleventy.js` and `UserConfig.js` but for the short term I’m going to refactor the deps of `UserConfig.js` to see if we can make some progress here that makes everyone happy and makes the Configuration API more usable.

I’ve filed the Eleventy.js work at #3383.

**Comment by zachleat** [comment]
Alright I have a bit more clarity on this now.

The previous `index.d.ts` file we had on this project would not and did not pass a `tsc --strict` check because the project does not and did not include first-party `.d.ts` files. The `index.d.ts` file we had pointed to a JavaScript file and no other `.d.ts` files existed. From my testing it looks like `tsc --strict` throws an error if a module’s `.d.ts` file cannot be found.

```
node_modules/@11ty/eleventy/src/index.d.ts:1:24 - error TS7016: Could not find a declaration file for module './UserConfig.js'. '/Users/zachleat/Temp/eleventy-3097-ts-strict/node_modules/@11ty/eleventy/src/UserConfig.js' implicitly has an 'any' type.

1 import UserConfig from "./UserConfig.js";
                         ~~~~~~~~~~~~~~~~~


Found 1 error in node_modules/@11ty/eleventy/src/index.d.ts:1
```

The good part of this exploration is that I’m well educated on the topic now and I’m not opposed to including `.d.ts` files on the project (at some point), but we’ll need to work through and fix the remainder of the TypeScript slop errors first. I’ve finished the ones for `UserConfig.js` at #3384 but I’ll want to finish the rest of them (see #3383) before we ship `.d.ts` files with the project (and that won’t be in time for 3.0, full disclosure).

We’ll keep the existing `@typedef` approach in place as it accommodates IDE autocomplete use cases that folks are already using. In my testing, restoring the previous `index.d.ts` did not offer any additional benefit for IDE autocomplete over the `@jsdoc` approach we already have in place (nor did it pass `tsc --strict` as already noted).

So, TL;DR I do plan to merge #3384 in service of the long term goal of shipping `.d.ts` type files with Eleventy (#3383), but `.d.ts` files will not ship as part of 3.0.

We’ll continue to recommend folks use https://www.11ty.dev/docs/config/#type-definitions for IDE autocomplete.

If I’ve missed something important here, please let me know!

**Comment by zachleat** [comment]
With a shout out to the approach @panoply is pushing for with #3296 that seems to be a bit cleaner pattern for configuration moving forward using a `defineConfig` export instead of a `/** @param */` jsdoc comment.

**Comment by zachleat** [comment]
@uncenter No, probably not. TypeScript can export `.d.ts` files from JSDoc which is the method I’d prefer moving forward here, as TypeScript becomes a progressive enhancement on top of the JavaScript code base (rather than the other way around, which has a shorter shelf life and more maintenance risk). Happy to revisit a larger discussion about types if and when they’ve shipped as part of JavaScript though!

**Comment by zachleat** [comment]
Just some clarification on the history here, since it’s been a bit all over the place and it’s worth documenting at the very least for future me (sorry folks, I’m more educated on TypeScript now 😭).

* `index.d.ts` was originally added in 2.0.0-canary.19 via #2091 https://github.com/11ty/eleventy/blob/v2.0.0-canary.19/package.json
* it was removed prior to the first 3.0 release 3.0.0-alpha.1 https://github.com/11ty/eleventy/pull/3060
* it was re-added in 3.0.0-alpha.11 https://github.com/11ty/eleventy/pull/3291
* it was removed again in 3.0.0-alpha.15 https://github.com/11ty/eleventy/pull/3349

**Comment by briancarbone** [comment]
> So, TL;DR I do plan to merge #3384 in service of the long term goal of shipping `.d.ts` type files with Eleventy (#3383), but `.d.ts` files will not ship as part of 3.0.

You _can_ (technically and unofficially) use TypeScript for the main configuration (e.g. `eleventy.config.ts`) by using the new type stripping feature of Node:

```NODE_OPTIONS=--experimental-strip-types npx @11ty/eleventy --config=eleventy.config.ts```

Essentially, the `.ts` extension provides a hint for the editor to use the TypeScript LSP and related highlighting; Eleventy just sees an ESM module because Node modifies it first.

Unfortunately, since the emitted `.d.ts` files were removed, it's not terribly ergonomic. You have to set both `allowJs` and `checkJs` to true _and_ increase the module depth (`maxNodeModuleJsDepth`) to a figure greater than 1. It does work, though.

**Comment by Ryuno-Ki** [comment]
*sneaks https://github.com/11ty/eleventy/pull/582 as the original PR in

**Comment by Ryuno-Ki** [comment]
What could be an option: Following node's lead and have a package in DefinitelyTyped.

For example, I spotted https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69010

Documentation: https://github.com/DefinitelyTyped/DefinitelyTyped#how-can-i-contribute

**Comment by noelforte** [comment]
Having a DefinitelyTyped package would be a good holdover, but in what I've learned so far about TS it's more terse if .d.ts files can be installed along with the package.

I'd say if it's possible for types to be maintained centrally without having to rely on another package, use a community solution like `11ty.ts`, or hand author `.d.ts` files onesself that would definitely be the best. Start with a generated file or work with what's already been done in `11ty.ts` and then tweak to meet 90% of cases.

I also saw another option get mentioned over on [#3296](https://github.com/11ty/eleventy/issues/3296#issuecomment-2256979639): expose a `defineConfig` identity function that handles the typings, which is a convention already popular in the JS community.

**Comment by panoply** [comment]
**FWIW**

11ty.ts is probably the most complete and well typed solution available (currently) in this eco. The `defineConfig` tactic is the most elegant (imo) and DefinitelyTyped cannot support this. 

There is also the matter of plugins which a lot of folks seem to forget. [11ty.ts](https://github.com/panoply/e11ty/tree/master/packages/11ty.ts) will auto-type plugins that expose declarations using a _typical_ syntactical structure of plugin parameters by cherry-picking index 1 of the argument order.

**Comment by murtuzaalisurti** [comment]
@panoply does 11ty.ts support 11ty v3?

**Comment by panoply** [comment]
I have only just upgraded locally. Even though it's lacking some of the v3 API features, it comes with vast majority of the API already cooked so v3 API is by default already largely supported minus some of the new methods available and deprecation notices.

If I get some time tomorrow I'll update and bring support.

**Comment by noelforte** [comment]
> set both allowJs and checkJs to true and increase the module depth (maxNodeModuleJsDepth) to a figure greater than 1

@briancarbone I never thanked you for this suggestion. I searched for weeks for a solution to get TypeScript to pick up Eleventy's JSDoc comments so I could use them in my plugin and this suggestion was what did it!

It might be possible to import the types directly without having to set up Typescript like this, I'm curious if subpath exports might solve this issue (#3460)?

**Comment by njfamirm** [comment]
The UserConfig not exported in @11ty/eleventy

**Comment by njfamirm** [comment]
tsc can generate types from jsdoc, but require to config them to generate `.d.ts` using:

```json
"declaration": true,
"emitDeclarationOnly": true,
```

**Comment by Ryuno-Ki** [comment]
> I'm do this in the local and all types is generated and work fine, need apply this to 11ty?

It was there before: https://github.com/11ty/eleventy/commit/b94c20effafbae04de31020a886d1844e668bb71

I recall having been mentioned in https://github.com/11ty/eleventy/pull/3060 - Type Declarations are an on/off-epic in this project.

**Comment by Ryuno-Ki** [comment]
Works for me :grimacing: 

![screenshot of a project](https://github.com/user-attachments/assets/6c5016e6-c5c0-43ba-b875-8ec82c05565e)

I can successfully jump to Type Definition in VSCodium.

**Comment by njfamirm** [comment]
> Works for me 😬

Strange!
I'm checking the code in the repository and don't see any reason why it should work
I don't know why it works for you 😁

**Comment by uncenter** [comment]
Fwiw @njfamirm looks like your website repository is on Eleventy v2? This issue is specifically for v3 types.

**Comment by njfamirm** [comment]
> Fwiw @njfamirm looks like your website repository is on Eleventy v2? This issue is specifically for v3 types.

No, I'm use v3

**Comment by 1oglop1** [comment]
@Ryuno-Ki I have it working but it also stopped working for some reason 🤷 

@njfamirm I managed to get it working via JSDoc like this

This is for `v3.0.0`
```js
/** @param {import("markdown-it").default} mdlib */
function amedMd(mdlib){
    mdlib.use() // makrdown-it completion here
}

/** 
 * @param {import("@11ty/eleventy/UserConfig")} eleventyConfig 
 */
export default function(eleventyConfig){
    
    eleventyConfig.amendLibrary("md",(mdlib)=>amedMd(mdlib) )
}
```

this is for `v3.0.1-alpha.1` due to update in `package.json`
```
/** @param {import('@11ty/eleventy/UserConfig').default} eleventyConfig*/
export default function(eleventyConfig){
    
    eleventyConfig.amendLibrary("md",(mdlib)=>amedMd(mdlib) )
}
```

I have been hesitating to switch to TS but learning APIs in JS is nothing but a waste of time, I really with JS had type annotations instead of a forever fork of TypeScript. 
The discovery of these magical JSDoc comments took me over an hour without previous experience.

**Comment by panoply** [comment]
I'll sort out v3 on 11ty.ts. Majority of the API is covered. DefineConfig solves all the nuances:

https://github.com/11ty/eleventy/issues/3296#issue-2316559634

**Comment by panoply** [comment]
FWIW relying on JSDoc types through comment annotations will be flaky. It requires more work from the TS Lang Server. 

I explored this at great lengths. While comment annotations suffice for a lot, complex typings need more context, this is best achieved through either declarations or defineConfig.
